Add Babel 8 migration docs (#2765)
babel/website: docs/v8-migration.md

| `@babel/plugin-proposal-dynamic-import` | `@babel/plugin-transform-dynamic-import` |
| `@babel/plugin-proposal-export-namespace-from` | `@babel/plugin-transform-export-namespace-from` |
| `@babel/plugin-proposal-json-strings` | `@babel/plugin-transform-json-strings` |
| `@babel/plugin-proposal-logical-assignment-operators` | `@babel/plugin-transform-logical-assignment-operators` |
| `@babel/plugin-proposal-nullish-coalescing-operator` | `@babel/plugin-transform-nullish-coalescing-operator` |
| `@babel/plugin-proposal-numeric-separator` | `@babel/plugin-transform-numeric-separator` |
| `@babel/plugin-proposal-object-rest-spread` | `@babel/plugin-transform-object-rest-spread` |
| `@babel/plugin-proposal-optional-catch-binding` | `@babel/plugin-transform-optional-catch-binding` |
| `@babel/plugin-proposal-optional-chaining` | `@babel/plugin-transform-optional-chaining` |
| `@babel/plugin-proposal-private-methods` | `@babel/plugin-transform-private-methods` |
| `@babel/plugin-proposal-private-property-in-object` | `@babel/plugin-transform-private-property-in-object` |
| `@babel/plugin-proposal-unicode-property-regex` | `@babel/plugin-transform-unicode-property-regex` |

> ## Package Discontinued

Reviewer: Maybe we should mention all the now unnecessary syntax plugins

  };
  ```
  Adapt the example above if you are using `@babel/plugin-transform-modules-amd` or `@babel/plugin-transform-modules-umd`. You can start the migration prior to Babel 8.0.

  If you are using `@babel/cli` and passing `--module-ids`, `--module-root` and `--module-id` from command line, please create a Babel config `babel.config.js` and specify options there.

### `@babel/preset-env` {#configuration-change-preset-env}

![medium](https://img.shields.io/badge/risk%20of%20breakage%3F-medium-yellow.svg)

- `includes` and `excludes` respect renamed package names ([#15576](https://github.com/babel/babel/pull/15576))

  **Migration**: If `includes` or `excludes` contain any plugins mentioned in the [Packages Renames section](./v8-migration.md#package-renames), change it to the new name. For example,

  ```diff title="babel.config.json"
  {
    "presets": [[
      "@babel/preset-env",
      {
  -     "includes": ["proposal-optional-chaining"]
  +     "includes": ["transform-optional-chaining"]
      }
    ]]
  }
  ```

![low](https://img.shields.io/badge/risk%20of%20breakage%3F-low-yellowgreen.svg)

- Remove `uglify` target ([#12594](https://github.com/babel/babel/pull/12594))

>   **Migration**: The `uglify` target had been deprecated since 7.0.0-beta.0, if you still need this, use the [`forceAllTransforms`](preset-env.md#forcealltransforms) option.

Reviewer: ```suggestion
  **Migration**: The `uglify` target had been deprecated since 7.0.0, if you still need this, use the [`forceAllTransforms`](preset-env.md#forcealltransforms) option.
```

Most people use only stable versions, no need to mention betas :)

  If your code needs to run in an environment which doesn't support object spread, you can either use `@babel/preset-env` (recommended) or `@babel/plugin-transform-object-rest-spread`. If you want to transpile `Object.assign` down, you also need to enable `@babel/plugin-transform-object-assign`.
  In Babel 7.7.0, you can opt-in this behavior by using the `useSpread` option.

- Type check input options ([#12460](https://github.com/babel/babel/pull/12460))

  **Migration**: The preset will also report invalid option names. Refer to the [docs](./preset-react.md#options) and ensure valid usage.

![low](https://img.shields.io/badge/risk%20of%20breakage%3F-low-yellowgreen.svg)

- Disallow `filter` option in automatic runtime ([#15068](https://github.com/babel/babel/pull/15068/commits/e2dd3be6e38b0254bc69a8e52c265214235829c6))

  **Migration**: The `filter` option can only be used with the `classic` runtime. If you have switched to `automatic` runtime, you can safely remove this option. Otherwise please specify `runtime: "classic"`.

### `@babel/preset-typescript` {#configuration-change-preset-ts}

![high](https://img.shields.io/badge/risk%20of%20breakage%3F-high-red.svg)

> - Remove `isTSX` and `allExtensions` option ([#14955](https://github.com/babel/babel/pull/14955))

Reviewer: ```suggestion
- Remove `isTSX` and `allExtensions` options ([#14955](https://github.com/babel/babel/pull/14955))
```

    ```diff title="babel.config.json"
    {
      "presets": [
        ["@babel/preset-react", { "runtime": "automatic" }],
    -   ["@babel/preset-typescript", { "allExtensions": true, "isTSX": true }]
    +   ["@babel/preset-typescript"]
      ]
    }
    ```

    If you want to transpile files other than `.tsx`, such as `.vue`, use `ignoreExtensions: true`:

    ```diff title="babel.config.js"
    {
      overrides: [{
        include: /\.vue$/,
        presets: [
          ['@babel/preset-typescript', {
    -       allExtensions: true, isTSX: true
    +       ignoreExtensions: true
          }]
        ]
      }]
    }
    ```

    If you want to preserve the JSX format but transpile the TypeScript part, use `ignoreExtensions: true` and add [`@babel/plugin-syntax-jsx`](./plugin-syntax-jsx.md) to `plugins`.

  - `isTSX: false` and `allExtensions: true`

    Use `ignoreExtensions: true`, see the example above.

  - `isTSX: false` and `allExtensions: false`

    You can safely remove them.

![medium](https://img.shields.io/badge/risk%20of%20breakage%3F-medium-yellow.svg)

- Remove `allowDeclareFields` option ([#12461](https://github.com/babel/babel/pull/12461))

>   **Migration**: Remove the option from your config. You will probably be fine with the new behaviour. Previously `allowDeclareFields` enables transforming the `declare` syntax introduced in TypeScript 3.7, in Babel 8 we support the syntax without such a flag. See also the [compilation changes](#compilation-change-jsx) section.

Reviewer: ```suggestion
  **Migration**: Remove the option from your config, since it's now enabled by default.  Previously `allowDeclareFields` enables transforming the `declare` syntax introduced in TypeScript 3.7, in Babel 8 we support the syntax without such a flag. See also the [compilation changes](#compilation-change-jsx) section.
```

### `@babel/plugin-transform-typescript` {#configuration-change-transform-ts}

![medium](https://img.shields.io/badge/risk%20of%20breakage%3F-medium-yellow.svg)

- Remove `allowDeclareFields` option ([#12461](https://github.com/babel/babel/pull/12461))

  **Migration**: Remove the option from your config.

### `@babel/plugin-syntax-typescript` {#configuration-change-syntax-ts}

![high](https://img.shields.io/badge/risk%20of%20breakage%3F-high-red.svg)

- Remove `isTSX` option ([#14955](https://github.com/babel/babel/pull/14955))

  **Migration**: If you are using `isTSX: true`, remove this option and add [`@babel/plugin-syntax-jsx`](./plugin-syntax-jsx.md) to `plugins`:

  ```diff title="babel.config.json
  {
    "plugins": [
  -   ["@babel/plugin-syntax-typescript", { "isTSX": true }]
  +   ["@babel/plugin-syntax-typescript"]
  +   ["@babel/plugin-syntax-jsx"]
    ]
  }
  ```

  If you are using `isTSX: false`, you can safely remove them.

### `@babel/preset-flow` {#configuration-change-preset-flow}

![medium](https://img.shields.io/badge/risk%20of%20breakage%3F-medium-yellow.svg)

- Remove `allowDeclareFields` option ([#12457](https://github.com/babel/babel/pull/12457))

>   **Migration**: Remove the option from your config. You will probably be fine with the new behaviour. Previously `allowDeclareFields` enables transforming the `declare` syntax introduced in Flow 0.120.0, in Babel 8 we support the syntax without such a flag. See also the [compilation changes](#compilation-change-flow) section.

Reviewer: ```suggestion
  **Migration**: Remove the option from your config, since it's now enabled by default. Previously `allowDeclareFields` enables transforming the `declare` syntax introduced in Flow 0.120.0, in Babel 8 we support the syntax without such a flag. See also the [compilation changes](#compilation-change-flow) section.
```

- Type check input options ([#12460](https://github.com/babel/babel/pull/12460))

  **Migration**: The preset will also report invalid option names. Refer to the [docs](./preset-flow.md#options) and ensure valid usage.

### `@babel/plugin-transform-flow-strip-types` {#configuration-change-transform-flow}

![medium](https://img.shields.io/badge/risk%20of%20breakage%3F-medium-yellow.svg)

- Remove `allowDeclareFields` option ([#12457](https://github.com/babel/babel/pull/12457))

  **Migration**: Remove the option from your config. You will probably be fine with the new behaviour.

### `@babel/parser` {#configuration-change-parser}

![low](https://img.shields.io/badge/risk%20of%20breakage%3F-low-yellowgreen.svg)

- Remove `estree` plugin option `classFeatures` ([#13752](https://github.com/babel/babel/pull/13752))

>   **Migration**: Remove the option from your config. You will probably be fine with the new behaviour. Previously the `classFeatures` plugin enables `@babel/parser` to produce class properties AST compatible with ESLint 8. In Babel 8 the `eslint-parser` only works with ESLint 8 and above.

Reviewer: ```suggestion
  **Migration**: Remove the option from your config, since it's now enabled by default. Previously the `classFeatures` plugin enables `@babel/parser` to produce class properties AST compatible with ESLint 8, following the ESTree specification. In Babel 8 the `eslint-parser` only works with ESLint 8 and above.
```
